File: apps/api/app/schemas/analytics.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class ExecutiveKpiResponse(BaseModel):
    availabilityScore: float
    healthScore: float
    automationScore: float
    complianceScore: float
    riskScore: int
    totalDevices: int
    activeAlerts: int
    recentExecutions: int


class CapacityForecastSchema(BaseModel):
    resourceName: str
    currentUsagePct: float
    growthRateMonthlyPct: float
    estimatedExhaustionDays: int
    status: str


class ComplianceScorecardResponse(BaseModel):
    overallCompliancePct: float
    passwordPoliciesPct: float
    firmwareCompliancePct: float
    aclCompliancePct: float
    goldenBaselinePct: float
    totalViolations: int


class GenerateReportRequest(BaseModel):
    report_type: str = "EXECUTIVE_SUMMARY"
    export_format: str = "PDF"


class ReportArtifactResponse(BaseModel):
    artifact_id: str
    title: str
    report_type: str
    format: str
    file_url: str
    file_size_bytes: int
    generated_by: str
    created_at: str


class CreateScheduleRequest(BaseModel):
    report_type: str = "EXECUTIVE_SUMMARY"
    frequency: str = "WEEKLY"
    recipients: List[str] = ["noc-team@switchpilot.io"]
    format: str = "PDF"


class ReportScheduleResponse(BaseModel):
    schedule_id: str
    report_type: str
    frequency: str
    recipients: List[str]
    format: str
    is_active: bool
    next_run_at: str

File: packages/types/src/analytics.ts
export enum ReportType {
  EXECUTIVE_SUMMARY = 'EXECUTIVE_SUMMARY',
  NETWORK_HEALTH = 'NETWORK_HEALTH',
  DEVICE_INVENTORY = 'DEVICE_INVENTORY',
  CONFIG_DRIFT = 'CONFIG_DRIFT',
  COMPLIANCE = 'COMPLIANCE',
  SECURITY = 'SECURITY',
  FIRMWARE_LIFECYCLE = 'FIRMWARE_LIFECYCLE',
  CAPACITY_PLANNING = 'CAPACITY_PLANNING',
  AVAILABILITY = 'AVAILABILITY',
  PERFORMANCE = 'PERFORMANCE',
  AUTOMATION_ACTIVITY = 'AUTOMATION_ACTIVITY',
  EXECUTION_SUCCESS_RATE = 'EXECUTION_SUCCESS_RATE',
  ROLLBACK_STATISTICS = 'ROLLBACK_STATISTICS',
  ALERT_STATISTICS = 'ALERT_STATISTICS',
  AI_USAGE = 'AI_USAGE',
  LICENSE_USAGE = 'LICENSE_USAGE',
  AUDIT_REPORTS = 'AUDIT_REPORTS',
}

export enum ExportFormat {
  PDF = 'PDF',
  EXCEL = 'EXCEL',
  CSV = 'CSV',
  JSON = 'JSON',
}

export interface ExecutiveKpis {
  availabilityScore: number;
  healthScore: number;
  automationScore: number;
  complianceScore: number;
  riskScore: number;
  totalDevices: number;
  activeAlerts: number;
  recentExecutions: number;
}

export interface CapacityForecastItem {
  resourceName: string;
  currentUsagePct: number;
  growthRateMonthlyPct: number;
  estimatedExhaustionDays: number;
  status: 'OPTIMAL' | 'WARNING' | 'CRITICAL';
}

export interface ComplianceScorecard {
  overallCompliancePct: number;
  passwordPoliciesPct: number;
  firmwareCompliancePct: number;
  aclCompliancePct: number;
  goldenBaselinePct: number;
  totalViolations: number;
}

export interface ReportScheduleEntity {
  id: string;
  organizationId: string;
  reportType: ReportType;
  frequency: 'DAILY' | 'WEEKLY' | 'MONTHLY' | 'QUARTERLY' | 'YEARLY';
  recipients: string[];
  format: ExportFormat;
  isActive: boolean;
  lastRunAt?: string;
  nextRunAt?: string;
}

export interface ReportArtifactEntity {
  id: string;
  title: string;
  reportType: ReportType;
  format: ExportFormat;
  fileUrl?: string;
  fileSizeBytes: number;
  generatedBy: string;
  createdAt: string;
}
